internal/oci/runtime_oci.go

```diff
@@ -898,73 +899,89 @@ func (r *runtimeOCI) AttachContainer(c *Container, inputStream io.Reader, output
 	return nil
 }
 
-// PortForwardContainer forwards the specified port provides statistics of a container.
-func (r *runtimeOCI) PortForwardContainer(c *Container, port int32, stream io.ReadWriter) error {
-	emptyStreamOnError := true
-	defer func() {
-		if emptyStreamOnError {
-			go func() {
-				if _, copyError := pools.Copy(ioutil.Discard, stream); copyError != nil {
-					logrus.Errorf("error closing port forward stream after other error: %v", copyError)
-				}
-			}()
+// PortForwardContainer forwards the specified port into the provided container.
+func (r *runtimeOCI) PortForwardContainer(
+	ctx context.Context, c *Container, nsPath string,
+	port int32, stream io.ReadWriteCloser,
+) error {
+	log.Infof(ctx,
+		"Starting port forward for %s in network namespace %s", c.ID(), nsPath,
+	)
+
+	// Adapted reference implementation:
+	// https://github.com/containerd/cri/blob/8c366d/pkg/server/sandbox_portforward_unix.go#L65-L120
+	if err := ns.WithNetNSPath(nsPath, func(_ ns.NetNS) error {
+		defer stream.Close()
+
+		// TODO: hardcoded to tcp4 because localhost resolves to ::1 by default
+		// if the system has IPv6 enabled. Theoretically happy eyeballs will
+		// try IPv6 first and fallback to IPv4 but resolving localhost doesn't
+		// seem to return and IPv4 address, thus failing the connection.
+		conn, err := net.Dial("tcp4", fmt.Sprintf("localhost:%d", port))
```

Comment:
@danwinship @dcbw @squeed would this fail on ipv6 only nodes?

Reply:
I didn't dig more into this, it fails if localhost resolves to IPv6, at least in my tests with KIND, but we only have containerd in KIND. It works on IPv6 and IPv4 clusters this way.
I think that it depends on how the container is exposing the port inside the pod. 
It's common to still find some code in k/k hardcoded to 172.0.0.1 to avoid the `localhost` resolution ... :shrug:

Reply:
cri-o knows the pod's `eth0` IP doesn't it? Can't you just connect to that rather than to localhost?

Reply:
that won't work for hostNetwork true, right?
I think that communication has to be done always through localhost, this runs inside the same namespace,  however, I couldn't understand how the port is published inside the namespace and why it is always bound to 127.0.0.1 (at least in containerd)

Reply:
You don't need port-forwarding in the hostNetwork case though...



> why it is always bound to 127.0.0.1 (at least in containerd)

Virtually all servers running in containers just bind to `0.0.0.0`, so it doesn't matter if you redirect to `127.0.0.1` or their primary IP. Golang-based servers will automatically bind to both `0.0.0.0` and `::` if you try to bind to either, but software written in other languages may not do that so, if we always send to `127.0.0.1` then it will fail for _some_ single-stack IPv6 containers.

Reply:
I changed the implementation to choose the first sandbox IP.

Reply:
for completenes, it seems that was hardcoded for security reasons in the past
https://github.com/containerd/cri/pull/1470/files#r424526233